<table align="left">
  <td>
    <a href="https://colab.research.google.com/github/marco-canas/fundamentos_de_programacion/blob/main/2_clases/unidad1/2_chapter.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>
  </td>
  <td>
    <a target="_blank" href="https://kaggle.com/kernels/welcome?src=https://github.com/marco-canas/fundamentos_de_programacion/blob/main/2_clases/unidad1/2_chapter.ipynb"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" /></a>
  </td>
</table>

### [Video de apoyo a la lectura interactiva y experimental de este cuaderno](https://studio.youtube.com/video/GpJTyzxyPyE/edit)

### [Vínculo al programa del curso de Fundamentos de programación:](https://github.com/marco-canas/fundamentos_de_programacion/blob/main/1_programa/1_programa_curso/Fundamentos%20de%20Programaci%C3%B3n%20314026.pdf)



## Código del curso: 314029 

Un reconocimiento a mis estudiantes que han construido conmigo este saber pedagógico:

<img src = 'https://github.com/marco-canas/algebra_y_trigonometria/blob/main/7_images/5_clase_3_2025-08-12_hora_11_04.jpg?raw=true'> 



In [4]:
import pandas as pd
import numpy as np

# Cargar la lista de estudiantes desde el archivo CSV
path = 'C:/Users/marco/Documentos/docencia/groups_list/fundamentos_programacion_314029_35.xlsx'
df = pd.read_excel(path)
df.index = np.arange(1, len(df) + 1)  # Cambiar el índice para que empiece en 1
df.head(17)


,Cédula,Nombre,Email
1,1038111190,CANOLE PEÑATE SEBASTIAN,sebastian.canolep@udea.edu.co
2,1048073285,CAPACHERO AGAMEZ JUAN PABLO,juan.capachero@udea.edu.co
3,1032251814,CASTRO ERAZO VALERIA,v.castro1@udea.edu.co
4,1038111332,CAUSIL ACOSTA ANDREW DAVID,a.causil@udea.edu.co
5,1063277880,CLEMENTE APARICIO JULIANA,juliana.clemente@udea.edu.co
6,1066604944,CLEMENTE APARICIO SAMUEL,s.clemente@udea.edu.co
7,1066600241,DURANGO TIRADO DANIEL DARIO,daniel.durango1@udea.edu.co
8,1038648760,MARIN VILLARREAL PABLO ALEJANDRO,pablo.marin1@udea.edu.co
9,1040515084,MOSQUERA TOSCANO STEFHANY PATRICIA,stefhany.mosquera@udea.edu.co
10,1126603686,ORTEGA QUINTERO GUILLERMO ANDRES,guillermo.ortegaq@udea.edu.co


# CHAPTER 2 


## [Configuración del entorno para el desarrollo Jupyter de Anaconda](https://read.amazon.com/?asin=B0BBB56L4X&ref_=kwl_kr_iv_rec_1)


## Secuencia de enseñanza del cuaderno "2_chapter.ipynb" — Capítulo 2: Configuración del entorno de desarrollo



El cuaderno pertenece al curso **Fundamentos de Programación** (código 314029) del profesor Marco Cañas, y sigue esta progresión pedagógica:



**1. Apertura y contexto del curso**
- Enlaces a Colab/Kaggle y video de apoyo para la lectura interactiva.
- Vínculo al programa del curso.
- Declaración de los **principios docentes**: modelo de aula invertida, metodología ABPP (Aprendizaje Basado en Problemas y Proyectos), desarrollo del pensamiento computacional (Python, R, Markdown), ciencia abierta/datos públicos, y enfoque TRRS de Duval.



**2. Presentación del capítulo (CHAPTER 2)**
- Título: *Environment Setup for Development* (Configuración del entorno para el desarrollo), con referencia al libro guía.
- **Structure**: adelanta los temas — configuración del entorno Python (a través de Windows PowerShell)(https://www.python.org/downloads/release/python-3147/), instalación de Anaconda, configuración de Jupyter Notebook (https://www.softonic.com/descargar/anaconda/windows/post-descarga?dt=internalDownload&installerType=riseInstaller), y prueba del entorno.
- **Objectives**: al terminar el capítulo el estudiante podrá configurar el entorno local, ejecutar código Python y trabajar con Jupyter Notebook y para verificar la instalación de ambos.



**3. Desarrollo práctico paso a paso**
- **Descarga e instalación de Anaconda**: 11 pasos numerados (descarga del instalador, ejecución, aceptación de licencia, ruta de instalación, opciones, finalización).
- **Prueba de la instalación**: verificación de que Anaconda quedó correctamente instalado.
- **Prueba de Python en la terminal Anaconda Prompt**: 4 pasos (abrir la consola, invocar Python, ejecutar comandos simples, salir).
- **Ejecución y prueba de Jupyter Notebook**: pasos para abrir Jupyter desde Anaconda, crear un notebook nuevo y ejecutar una celda de prueba.



**4. Cierre del capítulo**
- **Conclusión** que resume lo aprendido.
- **Preguntas de repaso** (ej. cómo comprobar la versión de Python, cómo abrir la consola interactiva, qué es Jupyter y cómo iniciarlo desde Anaconda).
- Enlace a la **evaluación docente**.
- Enlace de **continuidad** al siguiente capítulo (Capítulo 3: tipos de datos y estructuras básicas).
- Sección de **Referentes** bibliográficos.



En resumen, la secuencia sigue el patrón: *contexto/principios → objetivos y estructura del capítulo → instalación guiada paso a paso (Anaconda → terminal → Jupyter) → verificación práctica → cierre con conclusión, autoevaluación y enlace al siguiente tema*, coherente con el modelo de aula invertida declarado al inicio.

[Presentación de la secuencia de enseñanza de esta clase](https://www.youtube.com/watch?v=qCTz5uCaUp0)

Este capítulo mostrará paso a paso cómo instalar el gestor de paquetes Anaconda y Jupyter Notebook para el desarrollo en Python en un equipo Windows para un proyecto de ciencia de datos. 

Necesitamos el software Python para la instalación; además, necesitamos instalar muchas otras bibliotecas específicas para la tarea. 

Para el análisis y la ciencia de datos, el proyecto [Anaconda](https://www.anaconda.com/download) es muy popular, ya que es fácil de instalar y usar. 

Anaconda es un gestor de paquetes robusto que incluye numerosos paquetes esenciales de código abierto preinstalados (Pandas, NumPy, Matplotlib, etc.). 

In [2]:
!python --version

Python 3.11.9


A lo largo de este libro, utilizaremos la versión 3.11.9 de Python y Jupyter Notebook.

# Structure 

En este capítulo, discutiremos los siguientes temas: 

* Configuración del entorno para el desarrollo de Python 
* Instalación de [Anaconda](https://www.anaconda.com/download) 
* Configuración de Jupyter IPython Notebook 
* Prueba del entorno o la prueba de que lo que instalamos funciona. 

# Objectives   

Después de estudiar este capítulo, podrá:
* Configurar el entorno de desarrollo de Python en el equipo local
* Trabajar con Jupyter Notebook
* Ejecutar código Python para probar la instalación

# Fundamentos de Programacion

## Capítulo 2

Estamos viendo *cómo se trabaja* con **cuaderno jupyter**. 

In [3]:
x = 17 
x 

17


# Downloading and installing the Anaconda package   

Aquí, tenemos los pasos de instalación de Anaconda en la máquina con Microsoft Windows 10. 

## Paso 1: 

Vaya a https://www.anaconda.com/distribution/#download-section, obtendrá la pantalla que se muestra a continuación y haga clic en el botón Descargar(Download Now).  



<img src = 'https://github.com/marco-canas/fundamentos_de_programacion/blob/main/images/anaconda_1.png?raw=true'>

## Paso 2:   

Una vez que haga clic en la página de descarga, comenzará a descargar el archivo .exe de instalación (Anaconda3-2025.12-1-Windows-x86_64).

<img src = 'https://github.com/marco-canas/fundamentos_de_programacion/blob/main/images/anaconda_descarga_2.png?raw=true'> 

En la captura de pantalla anterior, puede ver el inicio de la descarga del archivo .exe de Anaconda. 

## Paso 3:  

Una vez completada la descarga, haga clic derecho en el archivo de instalación (Anaconda3-2025.12-1-Windows-x86_64) y seleccione "Ejecutar como administrador".

<img src = 'https://github.com/marco-canas/fundamentos_de_programacion/blob/main/images/anaconta_ejecutar_como_administrador_3.png?raw=true'>

## Paso 4: 

Haga clic en el botón Siguiente, como se muestra en la siguiente captura de pantalla:  

<img src = ''>

<img src = 'https://github.com/marco-canas/fundamentos_de_programacion/blob/main/images/inicio_instalacion_anaconda_1.jpg?raw=true'>

## Paso 5: 

Haga clic en el botón Acepto después de leer el Acuerdo de licencia.  

<img src = ''>

<img src = ''>

## Paso 6: 

Haga clic en el botón Siguiente después de seleccionar la opción Solo yo/Todos los usuarios, como se muestra a continuación. En este caso, es Todos los usuarios.

<img src = 'https://github.com/marco-canas/fundamentos_de_programacion/blob/main/images/seleccion_tipo_instalacion_2.jpeg?raw=true'>

## Paso 7: 

Ahora, especifique la ruta de la carpeta de instalación y haga clic en el botón Siguiente.

<img src = 'https://github.com/marco-canas/fundamentos_de_programacion/blob/main/images/elegir_localizacion_instalacion_3.jpeg?raw=true'>

## Paso 8: 

Ahora, marque ambas casillas de verificación y haga clic en el botón Instalar.

<img src = 'https://github.com/marco-canas/fundamentos_de_programacion/blob/main/images/opciones_instalaci%C3%B3n_avanzada_4.jpeg?raw=true'>

## Paso 9: 

Tras hacer clic en el botón Instalar, comenzará la instalación. Verá las siguientes pantallas; espere a que finalice la instalación:

<img src = 'https://github.com/marco-canas/fundamentos_de_programacion/blob/main/images/completando_anaconda_prompt_8.jpeg?raw=true'>

## Paso 10: 

Una vez completado, haga clic en el botón Siguiente.

<img src = 'https://github.com/marco-canas/fundamentos_de_programacion/blob/main/images/instalacion_completa_6.jpeg?raw=true'>

## Paso 11: 

Haga clic en el botón Finalizar en la nueva pantalla. Anaconda se instaló correctamente.

Al hacer clic en el botón Finalizar, se abrirá una página web en el navegador con más información sobre el producto Anaconda, que puede ignorar. 

<img src = 'https://github.com/marco-canas/fundamentos_de_programacion/blob/main/images/anaconda_navegator_9.jpeg?raw=true'>

En este punto, hemos completado la instalación de Anaconda. 

Ahora es momento de probarla y comprender el entorno de desarrollo de Python y Anaconda.

# Probando la instalación.   



Tras completar la instalación de Anaconda, comprobaremos la configuración de Python y Jupyter Notebook. 

Para verificar la instalación, siga estos pasos:

# Prueba de Python en la terminal Anaconda Promt  



## Paso 1: 

Ve a Inicio (Parte inferior izquierda de tu computador) y escribe Anaconda Prompt para abrir para abrir la Terminal asociada a Anaconda.  

<img src = 'https://github.com/marco-canas/fundamentos_de_programacion/blob/main/images/terminal_anaconda_prompt.png?raw=true'>

## Paso 2: 

Para comprobar si Python está instalado, escriba: `Python –-version` en el símbolo del sistema y pulse la tecla Enter.   
Si obtiene un resultado como el de la siguiente captura de pantalla, significa que Python se instaló correctamente:



<img src = 'python_version_desde_anaconda_prompt.png'>

## Paso 3: 

Ahora, escribe "python" y pulsa Enter para iniciar el Interactive Python Shell. 

Obtendrás un resultado como el de la siguiente captura de pantalla:  



<img src = 'verificar_python_en_anaconda_prompt.png'>

## Paso 4: 

Ahora escriba `print('Fundamentos de programación')` y presione Intro para ejecutar la instrucción print. 

Si la instalación se realizó correctamente, obtendrá un resultado como el de la siguiente captura de pantalla:

<img src = 'imprimir_fundamentos_de_programacion.png'> 

## Paso 4: 

Para salir del shell interactivo de Python, escriba `quit()` (abandonar) y presione Enter.



<img src = 'https://github.com/marco-canas/fundamentos_de_programacion/blob/main/images/quit_8.png?raw=true'> 

Ya hemos visto cómo ejecutar el código Python con el intérprete interactivo de Python. 

Veamos cómo podemos usar Jupiter Notebook para ejecutarlo.

# Ejecución y prueba de Jupyter Notebook  

Jupyter Notebook es una plataforma popular para escribir y ejecutar código Python entre científicos y analistas de datos.

Esta sección del capítulo mostrará cómo ejecutar Jupyter Notebook y el código Python. 

## Paso 1: 



Primero, En la terminal Anaconda Prompt creemos un directorio de trabajo (una carpeta simple de Windows) escribiendo el siguiente comando en Anaconda Prompt: `mkdir fundamentos_de_programacion_con_python`   


<img src = 'crar_una_carpeta_con_mkdir.png'>

## Paso 2: 

Luego, cambia el directorio escribiendo `cd fundamentos_de_programacion_con_python`. 

## [Paso 3:](https://read.amazon.com/?asin=B0BBB56L4X&ref_=kwl_kr_iv_rec_1) 



Ahora, escriba `jupyter notebook` en este Anaconda Prompt y presione Enter.  



<img src = 'https://github.com/marco-canas/fundamentos_de_programacion/blob/main/images/jupyter_notebook_10.png?raw=true'>

Iniciará el servidor local y obtendrá una página web de Jupyter Notebook como se muestra a continuación.  

<img src = 'pagina_web_de_jupyter_notebook.png'>

## Paso 4: 

Cambia el clic en el nuevo botón desplegable en la parte superior derecha y selecciona Python3.

<img src = 'seleccion_kernel_python_3.png'>

## Paso 5: 

Verá una página como la de la siguiente captura de pantalla, donde cada fila se denomina celda.   

Podemos agregar y eliminar celdas usando la opción del menú Archivo.

<img src = ''>

## Paso 4: 

Ahora, escribiremos y ejecutaremos la instrucción de impresión de Python.   

Primero, escriba el código Python que se muestra a continuación en la celda y, para ejecutarlo, presione Mayús+Intro (o use la opción de menú).  

Se ejecutará el código y obtendrá lo siguiente: `print("Bienvenido al curso de Fundamentos de programación")`

<img src = 'print_en_celda_cuaderno.png'>

Si ha completado con éxito todos los pasos mencionados anteriormente, significa que ha instalado correctamente el paquete Anaconda para el desarrollo en Python.

## Conclusión:  



En este capítulo, instalamos y probamos el entorno de desarrollo Anaconda-Python. 

Existen muchos IDE disponibles para el desarrollo en Python en el mercado. 

La elección del IDE depende totalmente del desarrollador; depende de su conveniencia y preferencia. 

En general, la mayoría de los científicos y analistas de datos utilizan Jupyter Notebook para su desarrollo inicial. 

En el siguiente capítulo, aprenderemos los fundamentos de la programación en Python con ejemplos prácticos de codificación. 

## Preguntas:   

* ¿Qué es Anaconda?   
* Lista de algunos paquetes/bibliotecas preinstalados en Anaconda. 
* ¿Cómo comprobar la versión de Python instalada? 
* ¿Cómo abrir la consola interactiva de Python? 
* ¿Qué es Jupyter Notebook y cómo se puede iniciar mediante Anaconda Prompt?

### [Evaluamos al profesor Marco Cañas Aquí](https://forms.office.com/Pages/ResponsePage.aspx?id=IefhmYRxjkmK_7KtTlPBwkanXIs1i1FEujpsZgO6dXpUREJPV1kxUk1JV1ozTFJIQVNIQjY5WEY3US4u)

### Continue su aprendizaje en la siguiente clase a través del siguiente [vínculo](https://github.com/marco-canas/fundamentos_de_programacion/blob/main/clases/unidad1/3_chapter.ipynb)

[CAPÍTULO 3. Sección 3.1 Tipos de datos integrados y Estructuras de datos básicas](https://github.com/marco-canas/fundamentos_de_programacion/blob/main/2_clases/unidad1/3_chapter_section_1_tipos_datos_estructuras_incorporadas.ipynb)

## Referentes 

[Libro guía Fundamentos de programación: Data Analysis with Python: Introducing NumPy, Pandas, Matplotlib, and Essential Elements of Python Programming (English Edition) [Kindle Edition]](https://read.amazon.com/?asin=B0BBB56L4X&ref_=kwl_kr_iv_rec_1)

# 20 ejercicios para preparar el primer parcial  

Con base en lo que ambos cuadernos buscan desarrollar —**Capítulo 1** (fundamentos de Python: paradigmas, características del lenguaje, REPL, tipos de datos, NumPy y progresiones aritméticas) y **Capítulo 2** (instalación de Anaconda, uso de Jupyter Notebook y verificación del entorno)— Los siguientes son 20 ejercicios de tarea organizados por habilidad, evitando repetir literalmente los que ya están en los cuadernos.

## Configuración del entorno (Cap. 2)

**1. Bitácora de instalación**
Documente en un notebook llamado `bitacora_instalacion.ipynb` los 11 pasos de instalación de Anaconda que siguió, con una captura de pantalla (o descripción) de cada paso y qué haría distinto si tuviera que instalarlo en otro computador.

**2. Diagnóstico del entorno**
Desde Anaconda Prompt, ejecute `conda --version`, `python --version` y `jupyter --version`. Explique en Markdown, dentro de un notebook, para qué sirve cada una de estas tres herramientas y cómo se relacionan entre sí.

**3. Kernel y entornos virtuales**
Investigue qué es un *entorno virtual* en Anaconda (`conda create -n nombre python=3.x`). Cree uno, actívelo, y desde Jupyter verifique con `import sys; sys.executable` que efectivamente está usando ese entorno.



**4. Jupyter desde cero**
Sin usar Anaconda Navigator, abra Jupyter Notebook exclusivamente desde la terminal (Anaconda Prompt). Enumere los comandos exactos que usó y explique qué es el "servidor local" que se abre en el navegador.

**5. Modos de celda**
Cree un notebook con al menos 3 celdas de código y 3 celdas Markdown intercaladas. Explique la diferencia entre el modo *comando* (Esc) y el modo *edición* (Enter) en Jupyter, y qué atajos de teclado usó.

## Características y paradigmas de Python (Cap. 1)

**6. Python interpretado vs compilado**
Escriba un pequeño script con un error de sintaxis intencional en la línea 5 pero código válido antes. Ejecútelo y explique, con evidencia de la salida, por qué el hecho de que Python sea interpretado permite que las líneas anteriores sí se ejecuten (o no, según el caso) antes de fallar.

**7. Refactorización a los tres paradigmas**
Escriba un programa que calcule el área y el perímetro de un rectángulo. Impleméntelo primero de forma **procedimental**, luego **funcional** (usando funciones puras) y finalmente **orientada a objetos** (con una clase `Rectangulo`). Compare las tres versiones en una celda Markdown.

**8. Tipado dinámico en acción**
Cree una variable `x`, asígnele un entero, imprima su tipo, luego reasígnela con una cadena y después con una lista. Imprima el tipo en cada paso. Explique qué significa que Python sea de *tipado dinámico* con este ejemplo como evidencia.

**9. Casos de uso por sector**
Elija dos de los sectores mencionados en el capítulo (salud, finanzas, ciencia de datos, IA/ML) e investigue una librería específica de Python usada en cada uno. Explique brevemente para qué sirve cada librería.

**10. REPL guiado**
Diseñe, como si fuera usted el profesor, una secuencia de 5 comandos en el REPL de Python (terminal, no notebook) que le enseñe a un compañero qué es la introspección de objetos (`type()`, `dir()`, `help()`).

## Variables, operadores y tipos de datos

**11. Calculadora de propina y factura**
Dado un valor de consulta `valor_consulta = 85000` y un porcentaje de propina `propina = 10`, calcule el valor de la propina y el total a pagar. Muestre los resultados con `print()` usando f-strings.

**12. Conversión de tipos (casting)**
Tome la cadena `"25"` y la cadena `"3.5"`. Conviértalas a `int` y `float` respectivamente, súmelas, e imprima el resultado junto con su tipo. Explique qué pasaría si intentara sumar las cadenas sin convertirlas.

**13. Booleanos y comparaciones**
Defina dos variables numéricas y evalúe con operadores de comparación (`>`, `<`, `==`, `!=`) al menos 4 expresiones booleanas distintas. Explique con sus palabras qué tipo de dato retornan estos operadores.



## NumPy y progresiones aritméticas

**14. Progresión descendente**
Use `np.arange()` para construir la progresión descendente $20, 17, 14, 11, 8, 5$. Identifique explícitamente cuál es el valor del paso (`step`) y explique por qué debe ser negativo.

**15. De progresión a suma**
Genere la progresión aritmética $3, 6, 9, \dots, 30$ con `np.arange()`, y luego calcule la suma total de sus términos usando `np.sum()`. Compare el resultado con la fórmula de suma de una progresión aritmética ($S_n = \frac{n(a_1+a_n)}{2}$).

**16. Progresión inversa: hallar el paso**
Dado que una progresión inicia en 4, termina en 64 y tiene 16 términos, calcule manualmente (con lápiz y papel o código) cuál sería la diferencia común, y verifique su resultado generando la progresión con `np.linspace()`.

**17. Función `progresion_geometrica()`**
Inspirándose en la función `progresion()` construida en clase para progresiones aritméticas, escriba una función `progresion_geometrica(inicio, razon, n_terminos)` que retorne una progresión geométrica (ej. $2, 4, 8, 16, \dots$). Pruébela con tres ejemplos.



**18. Visualización de una progresión**
Genere una progresión aritmética con `np.linspace()` de 20 términos entre 0 y 100, y grafíquela con `matplotlib` usando puntos (`plt.scatter`). Añada título y etiquetas a los ejes.

## Integración y documentación

**19. Notebook completo end-to-end**
Cree un notebook nuevo desde Jupyter (Cap. 2) llamado `taller_integrador.ipynb` que: importe NumPy, defina una función de progresión aritmética, genere dos progresiones distintas, las grafique, y documente cada paso con celdas Markdown explicando el propósito del código.

**20. Informe de errores comunes**
Durante el desarrollo de los ejercicios anteriores, registre (en una celda Markdown) al menos 3 errores que haya cometido (de instalación, de sintaxis o de tipos), qué mensaje de error mostró Python o Jupyter, y cómo lo solucionó. Este ejercicio refuerza tanto el manejo del entorno como la lectura de errores en Python.



# 20 problemas de preparación del primer parcial con situaciones problema de la ciencia de datos.   

Los siguientes son 20 problemas aplicados a situaciones reales de ciencia de datos, diseñados para que integre lo aprendido en ambos cuadernos: configuración del entorno, tipos de datos, paradigmas de programación, y NumPy (progresiones con `linspace()` y `arange()`).

## Sensores y muestreo de datos

**1. Sensor de temperatura ambiental**
Un sensor climático registra la temperatura cada hora durante un día (24 mediciones), comenzando en 15°C y terminando en 30°C. Genere esa secuencia con `np.linspace()`, calcule el promedio, la temperatura máxima y mínima, e imprima un resumen con f-strings.



**2. Muestreo de un sensor de presión atmosférica**
Un sensor mide presión atmosférica cada 5 minutos empezando en 1013 hPa con incrementos de -0.5 hPa (la presión va bajando por una tormenta que se aproxima), durante 2 horas. Constrúyalo con `np.arange()` y determine en qué medición la presión cae por debajo de 1000 hPa.

**3. Diseño de un experimento de dosis-respuesta**
Un investigador en salud necesita probar 12 concentraciones de un medicamento igualmente espaciadas entre 0 mg y 500 mg. Genere las concentraciones con la función apropiada (justifique si usó `linspace()` o `arange()`) y explique por qué esa elección es la correcta para este caso.

## Finanzas y economía

**4. Simulación de tasas de interés**
Un banco quiere simular escenarios de tasa de interés anual desde 2% hasta 8% en pasos de 0.5%. Genere la progresión, calcule cuántos escenarios se generaron, y para cada tasa calcule el monto final de una inversión de $1,000,000 a un año usando interés simple.

**5. Calendario de pagos de un préstamo**
Un préstamo se paga en 24 cuotas mensuales. Genere un arreglo con los números de cuota (1 a 24) usando `np.arange()`, y otro arreglo con el saldo restante suponiendo que cada cuota abona partes iguales del capital de $12,000,000. Grafique el saldo restante contra el número de cuota.

**6. Clasificación de transacciones por tipo de dato**
Dado un registro de una transacción bancaria con: `monto = 250000.75`, `moneda = "COP"`, `aprobada = True`, `id_transaccion = 10245`, identifique el tipo de dato de cada variable con `type()` y explique por qué cada tipo es el adecuado para ese campo (¿por qué `id_transaccion` no debería ser `float`?).

## Salud y ciencias biológicas

**7. Curva de crecimiento poblacional bacteriano**
Un cultivo bacteriano se mide cada 30 minutos durante 6 horas (13 mediciones). Genere los tiempos de medición con `np.linspace()` y simule (con valores inventados razonables) el número de bacterias en cada punto. Grafique la curva de crecimiento.

**8. Registro de signos vitales de pacientes**
Diseñe un mini sistema (usando un diccionario o una clase, aplicando el paradigma que prefiera) que almacene, para 5 pacientes, su nombre (cadena), edad (entero), temperatura corporal (flotante) y si están en estado crítico (booleano). Recorra el registro e imprima solo los pacientes en estado crítico.

**9. Progresión de dosis decreciente**
Un tratamiento inicia con una dosis de 400 mg y disminuye 25 mg cada semana durante 10 semanas. Genere la progresión con la función adecuada, y determine en qué semana la dosis llegaría a cero o menos (aunque clínicamente no tendría sentido, es un ejercicio matemático).

## Ciencia de datos y análisis exploratorio

**10. Comparación de escalas de muestreo**
Un dataset de ventas necesita ser resampleado en dos escenarios: (a) exactamente 15 puntos entre el primer y el último día del mes, (b) un punto cada 2 días durante el mes. Resuelva ambos con la función de NumPy correspondiente y compare cuántos puntos genera cada enfoque.

**11. Normalización manual de datos (min-max)**
Dado el arreglo de temperaturas generado en el ejercicio 1, escriba una función `normalizar(arreglo)` que transforme los valores a una escala entre 0 y 1 usando la fórmula $x' = \frac{x - min}{max - min}$. Pruébela y verifique que el resultado esté correctamente escalado.

**12. Función reutilizable para bins de histograma**
Escriba una función `generar_bins(minimo, maximo, n_bins)` que retorne los límites de los intervalos (bins) para construir un histograma, usando `np.linspace()`. Pruébela con datos de edades de una encuesta entre 18 y 65 años, con 8 bins.

**13. Detección de outliers simples**
Genere un arreglo de 20 "mediciones de ingresos mensuales" con `np.linspace()` entre 1,000,000 y 8,000,000, pero modifique manualmente un valor para que sea un outlier extremo (ej. 50,000,000). Escriba código que identifique y reporte cuál valor es el outlier, comparándolo con el promedio y la desviación estándar (`np.mean()`, `np.std()`).

## Programación orientada a objetos aplicada a datos

**14. Clase `Sensor` reutilizable**
Diseñe una clase `Sensor` con atributos `nombre`, `unidad` y un método `generar_lecturas(inicio, fin, n)` que internamente use `np.linspace()` para simular lecturas. Cree dos instancias: un sensor de temperatura y uno de humedad, y genere lecturas para ambos.

**15. Clase `Paciente` con validación de tipos**
Cree una clase `Paciente` cuyo constructor reciba nombre, edad y peso, y que internamente valide con `isinstance()` que edad sea `int` y peso sea `float`, lanzando un mensaje de error si no lo son. Pruebe la clase con un caso válido y uno inválido.

## Programación funcional aplicada a datos

**16. Pipeline funcional de limpieza de datos**
Dado un arreglo de temperaturas con algunos valores negativos "erróneos" (ej. sensores que fallaron), escriba una serie de funciones puras encadenadas: una que filtre los valores negativos, otra que calcule el promedio del resultado, y otra que redondee el promedio a 2 decimales. Aplique las tres en secuencia sobre los datos.



**17. Comparación procedimental vs funcional en un mismo problema**
Dado un conjunto de calificaciones de estudiantes generado con `np.arange(0, 105, 5)` (simulando posibles notas de 0 a 100), calcule cuántas calificaciones son aprobatorias (≥60) primero de forma **procedimental** (con un ciclo `for` y un contador) y luego de forma **funcional** (con `filter()` y `len()`, o con operaciones vectorizadas de NumPy). Compare ambas soluciones.



## Investigación y comunicación de resultados en Jupyter

**18. Reporte reproducible de un experimento**
Simule un experimento de laboratorio: 10 mediciones de pH de una solución tomadas cada 10 minutos entre pH 4 y pH 7 (use la función de NumPy adecuada). Documente el experimento completo en un notebook con: introducción en Markdown, código de generación de datos, gráfica y conclusión, siguiendo el modelo de "cuaderno reproducible" visto en los referentes del curso.



**19. Función `resumen_estadistico()`**
Escriba una función que reciba cualquier arreglo de NumPy y retorne (o imprima) un resumen con: mínimo, máximo, promedio, mediana y desviación estándar. Documéntela con *docstring* siguiendo el estilo del cuaderno del Capítulo 1, y pruébela con al menos tres conjuntos de datos distintos (temperaturas, ingresos, edades).



**20. Proyecto integrador: de la instalación al análisis**
Partiendo de cero (como si acabara de instalar Anaconda según el Cap. 2), cree un notebook nuevo que: (a) importe NumPy y matplotlib, (b) genere una progresión aritmética que represente una variable de su elección relacionada con ciencia de datos (ventas, temperatura, población, etc.), (c) implemente al menos una función propia para procesarla, (d) grafique el resultado, y (e) cierre con una sección Markdown de "Conclusiones", replicando la estructura pedagógica completa que han visto en ambos capítulos.


